plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -596,27 +601,152 @@ public boolean rename(Path src, Path dst)
     public boolean delete(Path path, boolean recursive)
             throws IOException
     {
-        try {
-            if (!directory(path)) {
-                return deleteObject(keyFromPath(path));
+        String bucketName = getBucketName(uri);
+        String key = keyFromPath(path);
+        ObjectMetadata s3ObjectMetadata = getS3ObjectMetadata(path, bucketName, key);
+        ObjectMetadata s3DirectoryMetadata = null;
+        if (!key.isEmpty()) {
+            s3DirectoryMetadata = getS3ObjectMetadata(path, bucketName, key + PATH_SEPARATOR);
+        }
+
+        if (recursive) {
+            boolean deleted;
+            if (key.endsWith(DIRECTORY_SUFFIX)) {
+                Path strippedPath = new Path(stripDirectorySuffix(path.toString()));
+                deleted = deletePrefix(strippedPath);
+                deleted = deleteObject(stripDirectorySuffix(key) + PATH_SEPARATOR) || deleted;
+            }
+            else {
+                deleted = deletePrefix(path);
+            }
+            if (s3ObjectMetadata != null) {
+                deleted = deleteObject(key) || deleted;
             }
+            if (s3DirectoryMetadata != null) {
+                String directoryKey = key + PATH_SEPARATOR;
+                deleted = deleteObject(directoryKey) || deleted;
```

Comment:
Might make the logic slightly less brittle to use `s3DirectoryMetadata.getKey()` so that we aren't reconstructing the key via concatenation in the original lookup and subsequent delete.